apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ProfileTaskExecutionContext.java

```diff
@@ -30,22 +37,139 @@
+    public void stopProfiling() {
+        if (profilingFuture != null) {
+            profilingFuture.cancel(true);
+        }
+    }
+
+    /**
+     * check have available slot to profile and add it
+     * @param tracingContext
+     * @param firstSpanOPName
+     * @return
+     */
+    public boolean attemptProfiling(TracingContext tracingContext, ID traceSegmentId, String firstSpanOPName) {
+        // check has available slot
+        final int usingSlotCount = currentProfilingCount.get();
+        if (usingSlotCount >= Config.Profile.MAX_PARALLEL) {
+            return false;
+        }
+
+        // check first operation name matches
+        if (!Objects.equals(task.getFistSpanOPName(), firstSpanOPName)) {
+            return false;
+        }
+
+        return tryToAttemptProfiling(tracingContext, traceSegmentId, firstSpanOPName, usingSlotCount);
+    }
+
+
+    /**
+     * profiling recheck
+     * @param tracingContext
+     * @param traceSegmentId
+     * @param firstSpanOPName
+     * @return
+     */
+    public boolean profilingRecheck(TracingContext tracingContext, ID traceSegmentId, String firstSpanOPName) {
+        boolean alreadyProfiling = tracingContext.isProfiling();
+
+        // not profiling and not available slot don't check anymore
+        int usingSlotCount = currentProfilingCount.get();
+        if (!alreadyProfiling && usingSlotCount >= Config.Profile.MAX_PARALLEL) {
+            return false;
+        }
+
+        // check first operation name matches
+        if (!Objects.equals(task.getFistSpanOPName(), firstSpanOPName)) {
+            if (alreadyProfiling) {
+                if (stopTracingProfile(tracingContext)) {
+                    // reduce total started profiling count when status is profiling
+                    totalStartedProfilingCount.addAndGet(-1);
+                }
+            }
+            return false;
+        } else if (alreadyProfiling) {
+            return true;
+        }
+
+        // not profiling, try to occupy slot
+        return tryToAttemptProfiling(tracingContext, traceSegmentId, firstSpanOPName, usingSlotCount);
+    }
+
+    /**
+     * find tracing context and clear on slot
+     *
+     * @param tracingContext
+     *
+     * @return current profiler is already start profiling
+     */
+    public boolean stopTracingProfile(TracingContext tracingContext) {
+        // find current tracingContext and clear it
+        boolean isProfilingStarted = false;
+        for (int slot = 0; slot < profilingSegmentSlots.length; slot++) {
+            ThreadProfiler currentProfiler = profilingSegmentSlots[slot];
+            if (currentProfiler != null && currentProfiler.matches(tracingContext)) {
+                profilingSegmentSlots[slot] = null;
+
+                // setting stop running
+                isProfilingStarted = currentProfiler.stopProfiling();
+                currentProfilingCount.addAndGet(-1);
+
+                profilingSegmentSlots = profilingSegmentSlots;
```

Comment:
You need to follow prefer review comments :) This is for activating `volatile` :P

Reply:
You are changing the element of an array, the volatile would work unless we do this.

Reply:
> You need to follow prefer review comments :) This is for activating `volatile` :P
> You are changing the element of an array, the volatile would work unless we do this.

Well sorry I indeed lose the context here, but why not simply use `java.util.concurrent.atomic.AtomicReferenceArray`, it does the same thing, even if we cannot use this class, for some reasons, it should be documented, later contributors may be hard to follow the review comments, especially when this gets merged, and people like you and me may forget the reason in some day, and the IDE is "too enthusiastic“ to hint that one should remove it:

<img width="792" alt="image" src="https://user-images.githubusercontent.com/15965696/72624365-c82a4800-3981-11ea-9a3b-0e93fca629e4.png">

Reply:
I remember that did this in different way. Can't remember the detail.Personally, I just use the easy way. 
 @mrproliu you could add doc, but don't use the link to that blog.

Reply:
> I just use the easy way.

IMHO, your meaning by saying "easy" is the data structure (array) itself, but in this case, we're tackling the thread-visibilities, I'd say it would be much more **straightforward** to use `java.util.concurrent.atomic.AtomicReferenceArray` than assigning the field to it self, Java programmers should be familiar with the former after all, my personal thought;

And you just brought me the interest to do a benchmark on the two solutions, in terms of the performance, check the following benchmark and results:

### Benchmark Codes
```java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.apache.skywalking.apm.plugin.dubbo;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReferenceArray;
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Warmup;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

@BenchmarkMode(Mode.Throughput)
@OutputTimeUnit(TimeUnit.MILLISECONDS)
@State(Scope.Thread)
@Fork(2)
@Warmup(iterations = 4)
@Measurement(iterations = 5)
public class AtomicArrayBenchmarkTest {
    final int arraySize = 50;
    final AtomicReferenceArray<Object> atomicReferenceArray = new AtomicReferenceArray<>(arraySize);

    @Benchmark
    public void testAtomicArray() {
        for (int i = 0; i < arraySize; i++) {
            atomicReferenceArray.set(i, new Object());
        }
    }

    volatile Object[] volatileArray = new Object[arraySize];

    @Benchmark
    public void testRawArray() {
        for (int i = 0; i < arraySize; i++) {
            volatileArray[i] = new Object();
            volatileArray = volatileArray;
        }
    }

    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.MICROSECONDS)
    public static void main(String[] args) throws RunnerException {
        Options options = new OptionsBuilder().include(AtomicArrayBenchmarkTest.class.getSimpleName()).build();
        new Runner(options).run();
    }
}

```

### Result

```text
Benchmark                                                Mode  Samples     Score  Score error   Units
o.a.s.a.p.d.AtomicArrayBenchmarkTest.testAtomicArray    thrpt       25  3003.470       37.575  ops/ms
o.a.s.a.p.d.AtomicArrayBenchmarkTest.testRawArray       thrpt       25  2766.009       27.934  ops/ms
```

it may be not that precise, but I'm with it :) since you said

> Profiling is happening in a high payload scenario

performance matters

Reply:
Make sense for switching to `AtomicReferenceArray` in this case.

Reply:
Change slot to the `AtomicReferenceArray`, please recheck. @wu-sheng @kezhenxu94